modules/fedimint-mint-client/src/backup/recovery.rs

```diff
@@ -0,0 +1,775 @@
+    spendable_note_by_nonce: BTreeMap<Nonce, (Amount, SpendableNote)>,
+
+    /// Outputs (by `OutPoint`) we track federation member confirmations for
+    /// blind nonces.
+    ///
+    /// Once we get enough confirmation (valid shares), these become new
+    /// spendable notes.
+    ///
+    /// Note that `NoteIssuanceRequest` is optional, as sometimes we might need
+    /// to handle a tx where only some of the blind nonces were in the pool.
+    /// A `None` means that this blind nonce/message is there only for
+    /// validation purposes, and will actually not create a
+    /// `spendable_note_by_nonce`
+    #[allow(clippy::type_complexity)]
+    pending_outputs: BTreeMap<
+        OutPoint,
+        (
+            TieredMulti<(CompressedBlindedMessage, Option<NoteIssuanceRequest>)>,
+            BTreeMap<PeerId, Vec<tbs::BlindedSignatureShare>>,
+        ),
+    >,
+
+    /// Next nonces that we expect might soon get used.
+    /// Once we see them, we move the tracking to `pending_outputs`
+    ///
+    /// Note: since looking up nonces is going to be the most common operation
+    /// the pool is kept shared (so only one lookup is enough), and
+    /// replenishment is done each time a note is consumed.
+    pending_nonces: BTreeMap<CompressedBlindedMessage, (NoteIssuanceRequest, NoteIndex, Amount)>,
+
+    /// Tail of `pending`. `pending_notes` is filled by generating note with
+    /// this index and incrementing it.
+    next_pending_note_idx: Tiered<NoteIndex>,
+
+    /// `LastECashNoteIndex` but tracked in flight. Basically max index of any
+    /// note that got a partial sig from the federation (initialled from the
+    /// backup value). TODO: One could imagine a case where the note was
+    /// issued but not get any partial sigs yet. Very unlikely in real life
+    /// scenario, but worth considering.
+    last_mined_nonce_idx: Tiered<NoteIndex>,
+
+    /// Threshold
+    threshold: u64,
+
+    /// Public key shares for each peer
+    ///
+    /// Used to validate contributed consensus items
+    pub_key_shares: BTreeMap<PeerId, Tiered<PublicKeyShare>>,
+
+    /// Aggregate public key for each amount tier
+    tbs_pks: Tiered<AggregatePublicKey>,
+
+    /// The number of nonces we look-ahead when looking for mints (per each
+    /// amount).
+    gap_limit: u64,
+}
+
+impl MintRestoreInProgressState {
+    fn transitions(
+        &self,
+        operation_id: OperationId,
+        context: &MintClientContext,
+        global_context: &DynGlobalClientContext,
+    ) -> Vec<StateTransition<MintRestoreStateMachine>> {
+        let global_context = global_context.clone();
+        let global_context_2 = global_context.clone();
+        let secret = context.secret.clone();
+        let self_clone = self.clone();
+        vec![StateTransition::new(
+            async move {
+                self_clone
+                    .make_progress(
+                        global_context.module_api(),
```

Comment:
This is the wrong api. Should be `global_context.api()`. This will send `module_1_fetch_epoch_history` RPC calls. It should send `fetch_epoch_history` RPC calls.

Reply:
Fixed

Reply:
Although it makes me wonder why CI didn't fail …

Reply:
There are no tests, and I rebased and fixed merged conflicts, and thought that `module_api` is the same thing I had `api()`.